pyphare/pyphare_tests/test_pharesee/test_geometry_2d.py

```diff
@@ -0,0 +1,493 @@
+                    {"boxes": [Box([27, 27], [27, 40])]},
+                    {"boxes": [Box([40, 27], [40, 40])]},
+                    {"boxes": [Box([28, 27], [39, 27])]},  # verify
+                    {"boxes": [Box([28, 40], [39, 40])]},  # verify
+                ]
+            },
+        ),
+    )
+    @unpack
+    def test_particle_level_ghost_boxes(self, expected):
+        print("GeometryTest.test_particle_level_ghost_boxes")
+        hierarchy = super().basic_hierarchy()
+
+        lvl_gaboxes = level_ghost_boxes(hierarchy, "particles")
+        for ilvl in range(1, len(hierarchy.patch_levels)):
+            self.assertEqual(len(lvl_gaboxes[ilvl].keys()), 1)
+
+            key = list(lvl_gaboxes[ilvl].keys())[0]
+
+            self.assertEqual(len(expected[ilvl]), len(lvl_gaboxes[ilvl][key]))
+
+            for actual, exp in zip(lvl_gaboxes[ilvl][key], expected[ilvl]):
+
+                for act_box, exp_box in zip(actual["boxes"], exp["boxes"]):
+                    self.assertEqual(act_box, exp_box)
+
+
+
+    def test_particle_level_ghost_boxes_do_not_overlap_patch_interiors(self):
+        print("GeometryTest.test_particle_level_ghost_boxes_do_not_overlap_patch_interiors")
+        hierarchy = super().basic_hierarchy()
+
+        lvl_gboxes = level_ghost_boxes(hierarchy, "particles")
+
+        assert len(lvl_gboxes) > 0
+        for ilvl, pdatainfos in lvl_gboxes.items():
+            assert len(pdatainfos) > 0
+            for particles_id, gaboxes_list in pdatainfos.items():
+                assert len(gaboxes_list) > 0
+                for pdatainfo in gaboxes_list:
+                    for box in pdatainfo["boxes"]:
+                        for patch in hierarchy.patch_levels[ilvl].patches:
+                            self.assertIsNone(patch.box * box)
+
+
+
+@ddt
+class ParticleLevelGhostGeometryTest(AGeometryTest):
+    @data(
+        (
+            {"L0": [Box2D(5, 9), Box2D(14, 19)]},
+            {
+                1: [
+                    Box([9, 9], [9, 20]),
+                    Box([20, 9], [20, 20]),
+                    Box([10, 9], [19, 9]),
+                    Box([10, 20], [19, 20]),
+                    Box([27, 27], [27, 40]),
+                    Box([40, 27], [40, 40]),
+                    Box([28, 27], [39, 27]),
+                    Box([28, 40], [39, 40]),
+                ]
+            },
+        ),
+        (
+            {"L0": [Box2D(5, 9), Box2D(10, 11)]},
+            {
+                1: [
+                    Box([9, 9], [9, 20]),
+                    Box([20, 9], [20, 19]),
+                    Box([10, 9], [19, 9]),
+                    Box([10, 20], [19, 20]),
+                    Box([19, 20], [19, 24]),
+                    Box([24, 19], [24, 24]),
+                    Box([20, 19], [23, 19]),
+                    Box([20, 24], [23, 24]),
+                ]
+            },
+        ),
+        (
+            {"L0": [Box2D(2, 9), Box([10, 1], [14, 10])]},
```

Comment:
we could add a little comment to each test case to explain quickly in what way it is specific

Reply:
![image](https://user-images.githubusercontent.com/30491/102475574-4e4f8280-405a-11eb-890c-67eae588df4d.png)